src/datastore/postgres-store.ts

```diff
@@ -5099,33 +5128,48 @@ export class PgDataStore
         atSingleBlock = false;
         queryParams.push(blockHeight);
       }
+      const principal = isValidPrincipal(args.stxAddress);
+      if (!principal) {
+        return { results: [], total: 0 };
+      }
       const resultQuery = await client.query<TxQueryResult & { count: number }>(
-        `
-        WITH principal_txs AS (
-          WITH event_txs AS (
-            SELECT tx_id FROM stx_events WHERE stx_events.sender = $1 OR stx_events.recipient = $1
-          )
-          SELECT *
-          FROM txs
-          WHERE canonical = true AND microblock_canonical = true AND (
-            sender_address = $1 OR
-            token_transfer_recipient_address = $1 OR
-            contract_call_contract_id = $1 OR
-            smart_contract_contract_id = $1
-          )
-          UNION
-          SELECT txs.* FROM txs
-          INNER JOIN event_txs
-          ON txs.tx_id = event_txs.tx_id
-          WHERE txs.canonical = true AND txs.microblock_canonical = true
-        )
-        SELECT ${TX_COLUMNS}, (COUNT(*) OVER())::integer as count
-        FROM principal_txs
-        ${atSingleBlock ? 'WHERE block_height = $4' : 'WHERE block_height <= $4'}
-        ORDER BY block_height DESC, microblock_sequence DESC, tx_index DESC
-        LIMIT $2
-        OFFSET $3
-        `,
+        // Smart contracts with a very high tx volume get frequent requests for the last N tx, where N
+        // is commonly <= 50. We'll query a materialized view if this is the case.
+        principal.type == 'contractAddress' && !atSingleBlock && args.limit + args.offset <= 50
+          ? `
```

Comment:
Nit: my preference here would be to break this out of a ternary expression and have an explicit `useMaterializedView: boolean`